// workers-ai-provider/dist/index.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);

// src/convert-to-workersai-chat-messages.ts
function convertToWorkersAIChatMessages(prompt) {
  const messages = [];
  const images = [];
  for (const { role, content } of prompt) {
    switch (role) {
      case "system": {
        messages.push({ content, role: "system" });
        break;
      }
      case "user": {
        const textParts = [];
        for (const part of content) {
          switch (part.type) {
            case "text": {
              textParts.push(part.text);
              break;
            }
            case "file": {
              if (part.data instanceof Uint8Array) {
                images.push({
                  image: part.data,
                  mediaType: part.mediaType,
                  providerOptions: part.providerOptions
                });
              }
              break;
            }
          }
        }
        messages.push({
          content: textParts.join("\n"),
          role: "user"
        });
        break;
      }
      case "assistant": {
        let text = "";
        const toolCalls = [];
        for (const part of content) {
          switch (part.type) {
            case "text": {
              text += part.text;
              break;
            }
            case "reasoning": {
              text += part.text;
              break;
            }
            case "file": {
              break;
            }
            case "tool-call": {
              toolCalls.push({
                function: {
                  arguments: JSON.stringify(part.input),
                  name: part.toolName
                },
                id: part.toolCallId,
                type: "function"
              });
              break;
            }
            case "tool-result": {
              break;
            }
            default: {
              const exhaustiveCheck = part;
              throw new Error(
                `Unsupported part type: ${exhaustiveCheck.type}`
              );
            }
          }
        }
        messages.push({
          content: text,
          role: "assistant",
          tool_calls: toolCalls.length > 0 ? toolCalls.map(({ function: { name, arguments: args }, id }) => ({
            function: { arguments: args, name },
            id,
            type: "function"
          })) : void 0
        });
        break;
      }
      case "tool": {
        for (const toolResponse of content) {
          if (toolResponse.type === "tool-result") {
            messages.push({
              content: JSON.stringify(toolResponse.output),
              name: toolResponse.toolName,
              tool_call_id: toolResponse.toolCallId,
              role: "tool"
            });
          }
        }
        break;
      }
      default: {
        const exhaustiveCheck = role;
        throw new Error(`Unsupported role: ${exhaustiveCheck}`);
      }
    }
  }
  return { images, messages };
}

// src/map-workersai-usage.ts
function mapWorkersAIUsage(output) {
  const usage = output.usage ?? {
    completion_tokens: 0,
    prompt_tokens: 0
  };
  const promptTokens = usage.prompt_tokens ?? 0;
  const completionTokens = usage.completion_tokens ?? 0;
  return {
    outputTokens: {
      total: completionTokens,
      text: void 0,
      reasoning: void 0
    },
    inputTokens: {
      total: promptTokens,
      noCache: void 0,
      cacheRead: void 0,
      cacheWrite: void 0
    },
    raw: { total: promptTokens + completionTokens }
  };
}

// src/streaming.ts
import { generateId } from "ai";

// src/map-workersai-finish-reason.ts
function mapWorkersAIFinishReason(finishReasonOrResponse) {
  let finishReason;
  if (typeof finishReasonOrResponse === "string" || finishReasonOrResponse === null || finishReasonOrResponse === void 0) {
    finishReason = finishReasonOrResponse;
  } else if (typeof finishReasonOrResponse === "object" && finishReasonOrResponse !== null) {
    const response = finishReasonOrResponse;
    const choices = response.choices;
    if (Array.isArray(choices) && choices.length > 0) {
      finishReason = choices[0].finish_reason;
    } else if ("finish_reason" in response) {
      finishReason = response.finish_reason;
    } else {
      finishReason = void 0;
    }
  } else {
    finishReason = void 0;
  }
  const raw = finishReason ?? "stop";
  switch (finishReason) {
    case "stop":
      return { unified: "stop", raw };
    case "length":
    case "model_length":
      return { unified: "length", raw };
    case "tool_calls":
      return { unified: "tool-calls", raw };
    case "error":
      return { unified: "error", raw };
    case "other":
    case "unknown":
      return { unified: "other", raw };
    default:
      return { unified: "stop", raw };
  }
}

// src/streaming.ts
function prependStreamStart(source, warnings) {
  let sentStart = false;
  return source.pipeThrough(
    new TransformStream({
      transform(chunk, controller) {
        if (!sentStart) {
          sentStart = true;
          controller.enqueue({
            type: "stream-start",
            warnings
          });
        }
        controller.enqueue(chunk);
      },
      flush(controller) {
        if (!sentStart) {
          controller.enqueue({
            type: "stream-start",
            warnings
          });
        }
      }
    })
  );
}
function isNullFinalizationChunk(tc) {
  const fn = tc.function;
  const name = fn?.name ?? tc.name ?? null;
  const args = fn?.arguments ?? tc.arguments ?? null;
  const id = tc.id ?? null;
  return !id && !name && (!args || args === "");
}
function getMappedStream(response) {
  const rawStream = response instanceof ReadableStream ? response : response.body;
  if (!rawStream) {
    throw new Error("No readable stream available for SSE parsing.");
  }
  let usage = {
    outputTokens: { total: 0, text: void 0, reasoning: void 0 },
    inputTokens: {
      total: 0,
      noCache: void 0,
      cacheRead: void 0,
      cacheWrite: void 0
    },
    raw: { totalTokens: 0 }
  };
  let textId = null;
  let reasoningId = null;
  let finishReason = null;
  let receivedDone = false;
  let receivedAnyData = false;
  const activeToolCalls = /* @__PURE__ */ new Map();
  const sseStream = rawStream.pipeThrough(new SSEDecoder());
  return sseStream.pipeThrough(
    new TransformStream({
      transform(data, controller) {
        if (!data || data === "[DONE]") {
          if (data === "[DONE]") receivedDone = true;
          return;
        }
        receivedAnyData = true;
        let chunk;
        try {
          chunk = JSON.parse(data);
        } catch {
          console.warn("[workers-ai-provider] failed to parse SSE event:", data);
          return;
        }
        if (chunk.usage) {
          usage = mapWorkersAIUsage(chunk);
        }
        const choices = chunk.choices;
        const choiceFinishReason = choices?.[0]?.finish_reason;
        const directFinishReason = chunk.finish_reason;
        if (choiceFinishReason != null) {
          finishReason = mapWorkersAIFinishReason(choiceFinishReason);
        } else if (directFinishReason != null) {
          finishReason = mapWorkersAIFinishReason(directFinishReason);
        }
        const nativeResponse = chunk.response;
        if (nativeResponse != null && nativeResponse !== "") {
          const responseText = String(nativeResponse);
          if (responseText.length > 0) {
            if (!textId) {
              textId = generateId();
              controller.enqueue({ type: "text-start", id: textId });
            }
            controller.enqueue({
              type: "text-delta",
              id: textId,
              delta: responseText
            });
          }
        }
        if (Array.isArray(chunk.tool_calls)) {
          emitToolCallDeltas(chunk.tool_calls, controller);
        }
        if (choices?.[0]?.delta) {
          const delta = choices[0].delta;
          const reasoningDelta = delta.reasoning_content;
          if (reasoningDelta && reasoningDelta.length > 0) {
            if (!reasoningId) {
              reasoningId = generateId();
              controller.enqueue({
                type: "reasoning-start",
                id: reasoningId
              });
            }
            controller.enqueue({
              type: "reasoning-delta",
              id: reasoningId,
              delta: reasoningDelta
            });
          }
          const textDelta = delta.content;
          if (textDelta && textDelta.length > 0) {
            if (!textId) {
              textId = generateId();
              controller.enqueue({ type: "text-start", id: textId });
            }
            controller.enqueue({
              type: "text-delta",
              id: textId,
              delta: textDelta
            });
          }
          const deltaToolCalls = delta.tool_calls;
          if (Array.isArray(deltaToolCalls)) {
            emitToolCallDeltas(deltaToolCalls, controller);
          }
        }
      },
      flush(controller) {
        for (const [, tc] of activeToolCalls) {
          controller.enqueue({ type: "tool-input-end", id: tc.id });
          controller.enqueue({
            type: "tool-call",
            toolCallId: tc.id,
            toolName: tc.toolName,
            input: tc.args
          });
        }
        if (reasoningId) {
          controller.enqueue({ type: "reasoning-end", id: reasoningId });
        }
        if (textId) {
          controller.enqueue({ type: "text-end", id: textId });
        }
        const effectiveFinishReason = !receivedDone && receivedAnyData && !finishReason ? {
          unified: "error",
          raw: "stream-truncated"
        } : finishReason ?? { unified: "stop", raw: "stop" };
        controller.enqueue({
          finishReason: effectiveFinishReason,
          type: "finish",
          usage
        });
      }
    })
  );
  function emitToolCallDeltas(toolCalls, controller) {
    for (const tc of toolCalls) {
      if (isNullFinalizationChunk(tc)) continue;
      const tcIndex = tc.index ?? 0;
      const fn = tc.function;
      const tcName = fn?.name ?? tc.name ?? null;
      const tcArgs = fn?.arguments ?? tc.arguments ?? null;
      const tcId = tc.id;
      if (!activeToolCalls.has(tcIndex)) {
        const id = tcId || generateId();
        const toolName = tcName || "";
        activeToolCalls.set(tcIndex, { id, toolName, args: "" });
        controller.enqueue({
          type: "tool-input-start",
          id,
          toolName
        });
        if (tcArgs != null && tcArgs !== "") {
          const delta = typeof tcArgs === "string" ? tcArgs : JSON.stringify(tcArgs);
          activeToolCalls.get(tcIndex).args += delta;
          controller.enqueue({
            type: "tool-input-delta",
            id,
            delta
          });
        }
      } else {
        const active = activeToolCalls.get(tcIndex);
        if (tcArgs != null && tcArgs !== "") {
          const delta = typeof tcArgs === "string" ? tcArgs : JSON.stringify(tcArgs);
          active.args += delta;
          controller.enqueue({
            type: "tool-input-delta",
            id: active.id,
            delta
          });
        }
      }
    }
  }
}
var SSEDecoder = class extends TransformStream {
  constructor() {
    let buffer = "";
    const decoder = new TextDecoder();
    super({
      transform(chunk, controller) {
        buffer += decoder.decode(chunk, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() || "";
        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed) continue;
          if (trimmed.startsWith("data: ")) {
            controller.enqueue(trimmed.slice(6));
          } else if (trimmed.startsWith("data:")) {
            controller.enqueue(trimmed.slice(5));
          }
        }
      },
      flush(controller) {
        if (buffer.trim()) {
          const trimmed = buffer.trim();
          if (trimmed.startsWith("data: ")) {
            controller.enqueue(trimmed.slice(6));
          } else if (trimmed.startsWith("data:")) {
            controller.enqueue(trimmed.slice(5));
          }
        }
      }
    });
  }
};

// src/utils.ts
import { generateId as generateId2 } from "ai";
function sanitizeToolCallId(id) {
  const alphanumeric = id.replace(/[^a-zA-Z0-9]/g, "");
  return alphanumeric.slice(0, 9).padEnd(9, "0");
}
function normalizeMessagesForBinding(messages) {
  return messages.map((msg) => {
    const normalized = { ...msg };
    if (normalized.content === null || normalized.content === void 0) {
      normalized.content = "";
    }
    if ("tool_call_id" in normalized && typeof normalized.tool_call_id === "string") {
      normalized.tool_call_id = sanitizeToolCallId(normalized.tool_call_id);
    }
    if ("tool_calls" in normalized && Array.isArray(normalized.tool_calls)) {
      normalized.tool_calls = normalized.tool_calls.map((tc) => ({
        ...tc,
        id: sanitizeToolCallId(tc.id)
      }));
    }
    return normalized;
  });
}
function createRun(config) {
  const { accountId, apiKey } = config;
  return async function run(model, inputs, options) {
    const {
      gateway: _gateway,
      prefix: _prefix,
      extraHeaders: _extraHeaders,
      returnRawResponse,
      signal,
      // AbortSignal — not serializable as a query parameter
      ...passthroughOptions
    } = options || {};
    const urlParams = new URLSearchParams();
    for (const [key, value] of Object.entries(passthroughOptions)) {
      if (value === void 0 || value === null) {
        throw new Error(
          `Value for option '${key}' is not able to be coerced into a string.`
        );
      }
      try {
        const valueStr = String(value);
        if (!valueStr) {
          continue;
        }
        urlParams.append(key, valueStr);
      } catch {
        throw new Error(
          `Value for option '${key}' is not able to be coerced into a string.`
        );
      }
    }
    const queryString = urlParams.toString();
    const url = `https://api.cloudflare.com/client/v4/accounts/${accountId}/ai/run/${model}${queryString ? `?${queryString}` : ""}`;
    const headers = {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json"
    };
    const body = JSON.stringify(inputs);
    const response = await fetch(url, {
      body,
      headers,
      method: "POST",
      signal
    });
    if (!response.ok && !returnRawResponse) {
      let errorBody;
      try {
        errorBody = await response.text();
      } catch {
        errorBody = "<unable to read response body>";
      }
      throw new Error(
        `Workers AI API error (${response.status} ${response.statusText}): ${errorBody}`
      );
    }
    if (returnRawResponse) {
      return response;
    }
    if (inputs.stream === true) {
      const contentType = response.headers.get("content-type") || "";
      if (contentType.includes("event-stream") && response.body) {
        return response.body;
      }
      if (response.body && !contentType.includes("json")) {
        return response.body;
      }
      const retryResponse = await fetch(url, {
        body: JSON.stringify({ ...inputs, stream: false }),
        headers,
        method: "POST",
        signal
      });
      if (!retryResponse.ok) {
        let errorBody;
        try {
          errorBody = await retryResponse.text();
        } catch {
          errorBody = "<unable to read response body>";
        }
        throw new Error(
          `Workers AI API error (${retryResponse.status} ${retryResponse.statusText}): ${errorBody}`
        );
      }
      const retryData = await retryResponse.json();
      return retryData.result;
    }
    const data = await response.json();
    return data.result;
  };
}
async function createRunBinary(config, model, audioBytes, contentType, signal) {
  const url = `https://api.cloudflare.com/client/v4/accounts/${config.accountId}/ai/run/${model}`;
  const response = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${config.apiKey}`,
      "Content-Type": contentType
    },
    body: audioBytes,
    signal
  });
  if (!response.ok) {
    let errorBody;
    try {
      errorBody = await response.text();
    } catch {
      errorBody = "<unable to read response body>";
    }
    throw new Error(
      `Workers AI API error (${response.status} ${response.statusText}): ${errorBody}`
    );
  }
  const data = await response.json();
  return data.result ?? data;
}
function prepareToolsAndToolChoice(tools, toolChoice) {
  if (tools == null) {
    return { tool_choice: void 0, tools: void 0 };
  }
  const mappedTools = tools.map((tool) => ({
    function: {
      description: tool.type === "function" && tool.description,
      name: tool.name,
      parameters: tool.type === "function" && tool.inputSchema
    },
    type: "function"
  }));
  if (toolChoice == null) {
    return { tool_choice: void 0, tools: mappedTools };
  }
  const type = toolChoice.type;
  switch (type) {
    case "auto":
      return { tool_choice: type, tools: mappedTools };
    case "none":
      return { tool_choice: type, tools: mappedTools };
    case "required":
      return { tool_choice: "any", tools: mappedTools };
    // Workers AI does not support tool mode directly,
    // so we filter the tools and force the tool choice through 'any'
    case "tool":
      return {
        tool_choice: "any",
        tools: mappedTools.filter((tool) => tool.function.name === toolChoice.toolName)
      };
    default: {
      const exhaustiveCheck = type;
      throw new Error(`Unsupported tool choice type: ${exhaustiveCheck}`);
    }
  }
}
function processToolCall(toolCall) {
  const fn = "function" in toolCall && typeof toolCall.function === "object" && toolCall.function ? toolCall.function : null;
  if (fn?.name) {
    return {
      input: typeof fn.arguments === "string" ? fn.arguments : JSON.stringify(fn.arguments || {}),
      toolCallId: toolCall.id || generateId2(),
      type: "tool-call",
      toolName: fn.name
    };
  }
  const flat = toolCall;
  return {
    input: typeof flat.arguments === "string" ? flat.arguments : JSON.stringify(flat.arguments || {}),
    toolCallId: flat.id || generateId2(),
    type: "tool-call",
    toolName: flat.name
  };
}
function processToolCalls(output) {
  if (output.tool_calls && Array.isArray(output.tool_calls)) {
    return output.tool_calls.map(
      (toolCall) => processToolCall(toolCall)
    );
  }
  const choices = output.choices;
  if (choices?.[0]?.message?.tool_calls && Array.isArray(choices[0].message.tool_calls)) {
    return choices[0].message.tool_calls.map((toolCall) => processToolCall(toolCall));
  }
  return [];
}
function processText(output) {
  const choices = output.choices;
  const choiceContent = choices?.[0]?.message?.content;
  if (choiceContent != null && String(choiceContent).length > 0) {
    return String(choiceContent);
  }
  if ("response" in output) {
    const response = output.response;
    if (typeof response === "object" && response !== null) {
      return JSON.stringify(response);
    }
    if (typeof response === "number") {
      return String(response);
    }
    if (response === null || response === void 0) {
      return void 0;
    }
    return String(response);
  }
  return void 0;
}

// src/aisearch-chat-language-model.ts
var AISearchChatLanguageModel = class {
  constructor(modelId, settings, config) {
    __publicField(this, "specificationVersion", "v3");
    __publicField(this, "defaultObjectGenerationMode", "json");
    __publicField(this, "supportedUrls", {});
    __publicField(this, "modelId");
    __publicField(this, "settings");
    __publicField(this, "config");
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
  }
  get provider() {
    return this.config.provider;
  }
  getWarnings({
    tools,
    frequencyPenalty,
    presencePenalty,
    responseFormat
  }) {
    const warnings = [];
    if (tools != null && tools.length > 0) {
      console.warn(
        "[workers-ai-provider] Tools are not supported by AI Search. They will be ignored."
      );
      warnings.push({ feature: "tools", type: "unsupported" });
    }
    if (frequencyPenalty != null) {
      warnings.push({ feature: "frequencyPenalty", type: "unsupported" });
    }
    if (presencePenalty != null) {
      warnings.push({ feature: "presencePenalty", type: "unsupported" });
    }
    if (responseFormat?.type === "json") {
      warnings.push({ feature: "responseFormat", type: "unsupported" });
    }
    return warnings;
  }
  /**
   * Build the search query from messages.
   * Flattens the conversation into a single string for aiSearch.
   */
  buildQuery(prompt) {
    const { messages } = convertToWorkersAIChatMessages(prompt);
    return messages.map(({ content, role }) => `${role}: ${content}`).join("\n\n");
  }
  async doGenerate(options) {
    const warnings = this.getWarnings(options);
    const query = this.buildQuery(options.prompt);
    const output = await this.config.binding.aiSearch({ query });
    return {
      finishReason: { unified: "stop", raw: "stop" },
      content: [
        ...output.data.map(({ file_id, filename, score }) => ({
          type: "source",
          sourceType: "url",
          id: file_id,
          url: filename,
          providerMetadata: {
            attributes: { score }
          }
        })),
        {
          type: "text",
          text: output.response
        },
        ...processToolCalls(output)
      ],
      usage: mapWorkersAIUsage(output),
      warnings
    };
  }
  async doStream(options) {
    const warnings = this.getWarnings(options);
    const query = this.buildQuery(options.prompt);
    const response = await this.config.binding.aiSearch({
      query,
      stream: true
    });
    return {
      stream: prependStreamStart(
        getMappedStream(response),
        warnings
      )
    };
  }
};

// src/workersai-embedding-model.ts
import { TooManyEmbeddingValuesForCallError } from "@ai-sdk/provider";
var WorkersAIEmbeddingModel = class {
  constructor(modelId, settings, config) {
    __publicField(this, "specificationVersion", "v3");
    __publicField(this, "modelId");
    __publicField(this, "config");
    __publicField(this, "settings");
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
  }
  get provider() {
    return this.config.provider;
  }
  get maxEmbeddingsPerCall() {
    return this.settings.maxEmbeddingsPerCall ?? 3e3;
  }
  get supportsParallelCalls() {
    return this.settings.supportsParallelCalls ?? true;
  }
  async doEmbed({ values }) {
    if (values.length > this.maxEmbeddingsPerCall) {
      throw new TooManyEmbeddingValuesForCallError({
        maxEmbeddingsPerCall: this.maxEmbeddingsPerCall,
        modelId: this.modelId,
        provider: this.provider,
        values
      });
    }
    const {
      gateway,
      maxEmbeddingsPerCall: _maxEmbeddingsPerCall,
      supportsParallelCalls: _supportsParallelCalls,
      ...passthroughOptions
    } = this.settings;
    const response = await this.config.binding.run(
      this.modelId,
      {
        text: values
      },
      {
        gateway: this.config.gateway ?? gateway,
        ...passthroughOptions
      }
    );
    return {
      embeddings: response.data,
      warnings: []
    };
  }
};

// src/workersai-chat-language-model.ts
import { generateId as generateId3 } from "ai";
var WorkersAIChatLanguageModel = class {
  constructor(modelId, settings, config) {
    __publicField(this, "specificationVersion", "v3");
    __publicField(this, "defaultObjectGenerationMode", "json");
    __publicField(this, "supportedUrls", {});
    __publicField(this, "modelId");
    __publicField(this, "settings");
    __publicField(this, "config");
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
  }
  get provider() {
    return this.config.provider;
  }
  getArgs({
    responseFormat,
    tools,
    toolChoice,
    maxOutputTokens,
    temperature,
    topP,
    frequencyPenalty,
    presencePenalty,
    seed
  }) {
    const type = responseFormat?.type ?? "text";
    const warnings = [];
    if (frequencyPenalty != null) {
      warnings.push({ feature: "frequencyPenalty", type: "unsupported" });
    }
    if (presencePenalty != null) {
      warnings.push({ feature: "presencePenalty", type: "unsupported" });
    }
    const baseArgs = {
      max_tokens: maxOutputTokens,
      model: this.modelId,
      random_seed: seed,
      safe_prompt: this.settings.safePrompt,
      temperature,
      top_p: topP
    };
    switch (type) {
      case "text": {
        return {
          args: {
            ...baseArgs,
            response_format: void 0,
            ...prepareToolsAndToolChoice(tools, toolChoice)
          },
          warnings
        };
      }
      case "json": {
        return {
          args: {
            ...baseArgs,
            response_format: {
              type: "json_schema",
              json_schema: responseFormat?.type === "json" ? responseFormat.schema : void 0
            },
            tools: void 0
          },
          warnings
        };
      }
      default: {
        const exhaustiveCheck = type;
        throw new Error(`Unsupported type: ${exhaustiveCheck}`);
      }
    }
  }
  /**
   * Build the inputs object for `binding.run()`, shared by doGenerate and doStream.
   */
  buildRunInputs(args, messages, images, options) {
    if (images.length > 1) {
      throw new Error("Multiple images are not yet supported as input");
    }
    const imagePart = images[0];
    const finalMessages = this.config.isBinding ? normalizeMessagesForBinding(messages) : messages;
    return {
      max_tokens: args.max_tokens,
      messages: finalMessages,
      temperature: args.temperature,
      tools: args.tools,
      top_p: args.top_p,
      ...imagePart ? { image: Array.from(imagePart.image) } : {},
      // Only include response_format when actually set
      ...args.response_format ? { response_format: args.response_format } : {},
      ...options?.stream ? { stream: true } : {}
    };
  }
  /**
   * Get passthrough options for binding.run() from settings.
   */
  getRunOptions() {
    const { gateway, safePrompt: _safePrompt, ...passthroughOptions } = this.settings;
    return {
      gateway: this.config.gateway ?? gateway,
      ...passthroughOptions
    };
  }
  async doGenerate(options) {
    const { args, warnings } = this.getArgs(options);
    const { messages, images } = convertToWorkersAIChatMessages(options.prompt);
    const inputs = this.buildRunInputs(args, messages, images);
    const runOptions = this.getRunOptions();
    const output = await this.config.binding.run(
      args.model,
      inputs,
      runOptions
    );
    if (output instanceof ReadableStream) {
      throw new Error(
        "Unexpected streaming response from non-streaming request. Check that `stream: true` was not passed."
      );
    }
    const outputRecord = output;
    const choices = outputRecord.choices;
    const reasoningContent = choices?.[0]?.message?.reasoning_content;
    return {
      finishReason: mapWorkersAIFinishReason(outputRecord),
      content: [
        ...reasoningContent ? [{ type: "reasoning", text: reasoningContent }] : [],
        {
          type: "text",
          text: processText(outputRecord) ?? ""
        },
        ...processToolCalls(outputRecord)
      ],
      usage: mapWorkersAIUsage(output),
      warnings
    };
  }
  async doStream(options) {
    const { args, warnings } = this.getArgs(options);
    const { messages, images } = convertToWorkersAIChatMessages(options.prompt);
    const inputs = this.buildRunInputs(args, messages, images, { stream: true });
    const runOptions = this.getRunOptions();
    const response = await this.config.binding.run(
      args.model,
      inputs,
      runOptions
    );
    if (response instanceof ReadableStream) {
      return {
        stream: prependStreamStart(getMappedStream(response), warnings)
      };
    }
    const outputRecord = response;
    const choices = outputRecord.choices;
    const reasoningContent = choices?.[0]?.message?.reasoning_content;
    let textId = null;
    let reasoningId = null;
    return {
      stream: new ReadableStream({
        start(controller) {
          controller.enqueue({
            type: "stream-start",
            warnings
          });
          if (reasoningContent) {
            reasoningId = generateId3();
            controller.enqueue({ type: "reasoning-start", id: reasoningId });
            controller.enqueue({
              type: "reasoning-delta",
              id: reasoningId,
              delta: reasoningContent
            });
            controller.enqueue({ type: "reasoning-end", id: reasoningId });
          }
          const text = processText(outputRecord);
          if (text) {
            textId = generateId3();
            controller.enqueue({ type: "text-start", id: textId });
            controller.enqueue({ type: "text-delta", id: textId, delta: text });
            controller.enqueue({ type: "text-end", id: textId });
          }
          for (const toolCall of processToolCalls(outputRecord)) {
            controller.enqueue(toolCall);
          }
          controller.enqueue({
            type: "finish",
            finishReason: mapWorkersAIFinishReason(outputRecord),
            usage: mapWorkersAIUsage(response)
          });
          controller.close();
        }
      })
    };
  }
};

// src/workersai-image-model.ts
var WorkersAIImageModel = class {
  constructor(modelId, settings, config) {
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
    __publicField(this, "specificationVersion", "v3");
  }
  get maxImagesPerCall() {
    return this.settings.maxImagesPerCall ?? 1;
  }
  get provider() {
    return this.config.provider;
  }
  async doGenerate({
    prompt,
    n,
    size,
    aspectRatio,
    seed
  }) {
    const { width, height } = getDimensionsFromSizeString(size);
    const warnings = [];
    if (aspectRatio != null) {
      warnings.push({
        details: "This model does not support aspect ratio. Use `size` instead.",
        feature: "aspectRatio",
        type: "unsupported"
      });
    }
    const generateImage = async () => {
      const output = await this.config.binding.run(this.modelId, {
        height,
        prompt: prompt ?? "",
        seed,
        width
      });
      return toUint8Array(output);
    };
    const images = await Promise.all(
      Array.from({ length: n }, () => generateImage())
    );
    return {
      images,
      response: {
        headers: {},
        modelId: this.modelId,
        timestamp: /* @__PURE__ */ new Date()
      },
      warnings
    };
  }
};
function getDimensionsFromSizeString(size) {
  const [width, height] = size?.split("x") ?? [void 0, void 0];
  return {
    height: parseInteger(height),
    width: parseInteger(width)
  };
}
function parseInteger(value) {
  if (value === "" || !value) return void 0;
  const number = Number(value);
  return Number.isInteger(number) ? number : void 0;
}
async function toUint8Array(output) {
  if (output instanceof Uint8Array) {
    return output;
  }
  if (output instanceof ArrayBuffer) {
    return new Uint8Array(output);
  }
  if (output instanceof ReadableStream) {
    const reader = output.getReader();
    const chunks = [];
    let totalLength = 0;
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      chunks.push(value);
      totalLength += value.length;
    }
    const result = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of chunks) {
      result.set(chunk, offset);
      offset += chunk.length;
    }
    return result;
  }
  if (output instanceof Response) {
    return new Uint8Array(await output.arrayBuffer());
  }
  if (typeof output === "object" && output !== null) {
    const obj = output;
    if (typeof obj.image === "string") {
      return Uint8Array.from(atob(obj.image), (c) => c.charCodeAt(0));
    }
    if (obj.data instanceof Uint8Array) {
      return obj.data;
    }
    if (obj.data instanceof ArrayBuffer) {
      return new Uint8Array(obj.data);
    }
    if (typeof obj.arrayBuffer === "function") {
      return new Uint8Array(await obj.arrayBuffer());
    }
  }
  throw new Error(
    `Unexpected output type from image model. Got ${Object.prototype.toString.call(output)} with keys: ${typeof output === "object" && output !== null ? JSON.stringify(Object.keys(output)) : "N/A"}`
  );
}

// src/workersai-transcription-model.ts
var WorkersAITranscriptionModel = class {
  constructor(modelId, settings, config) {
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
    __publicField(this, "specificationVersion", "v3");
  }
  get provider() {
    return this.config.provider;
  }
  async doGenerate(options) {
    const { audio, mediaType, abortSignal } = options;
    const warnings = [];
    const audioBytes = typeof audio === "string" ? Uint8Array.from(atob(audio), (c) => c.charCodeAt(0)) : audio;
    const isNova3 = this.modelId === "@cf/deepgram/nova-3";
    let rawResult;
    if (isNova3) {
      rawResult = await this.runNova3(audioBytes, mediaType, abortSignal);
    } else {
      rawResult = await this.runWhisper(audioBytes, abortSignal);
    }
    const result = rawResult;
    if (isNova3) {
      return this.normalizeNova3Response(result, warnings);
    }
    return this.normalizeWhisperResponse(result, warnings);
  }
  // ---------------------------------------------------------------------------
  // Whisper models
  // ---------------------------------------------------------------------------
  async runWhisper(audioBytes, abortSignal) {
    const modelStr = this.modelId;
    const audio = modelStr === "@cf/openai/whisper-large-v3-turbo" ? uint8ArrayToBase64(audioBytes) : Array.from(audioBytes);
    const inputs = { audio };
    if (this.settings.language) {
      inputs.language = this.settings.language;
    }
    if (this.settings.prompt) {
      inputs.initial_prompt = this.settings.prompt;
    }
    return this.config.binding.run(
      this.modelId,
      inputs,
      { gateway: this.config.gateway, signal: abortSignal }
    );
  }
  normalizeWhisperResponse(raw, warnings) {
    const text = raw.text ?? "";
    const segments = [];
    if (raw.segments && Array.isArray(raw.segments)) {
      for (const seg of raw.segments) {
        segments.push({
          text: seg.text ?? "",
          startSecond: seg.start ?? 0,
          endSecond: seg.end ?? 0
        });
      }
    } else if (raw.words && Array.isArray(raw.words)) {
      for (const w of raw.words) {
        segments.push({
          text: w.word ?? "",
          startSecond: w.start ?? 0,
          endSecond: w.end ?? 0
        });
      }
    }
    const info = raw.transcription_info;
    return {
      text,
      segments,
      language: info?.language ?? void 0,
      durationInSeconds: info?.duration ?? void 0,
      warnings,
      response: {
        timestamp: /* @__PURE__ */ new Date(),
        modelId: this.modelId,
        headers: {}
      }
    };
  }
  // ---------------------------------------------------------------------------
  // Deepgram Nova-3
  // ---------------------------------------------------------------------------
  async runNova3(audioBytes, mediaType, abortSignal) {
    if (this.config.isBinding) {
      return this.config.binding.run(
        this.modelId,
        {
          audio: { body: uint8ArrayToBase64(audioBytes), contentType: mediaType }
        },
        { gateway: this.config.gateway, signal: abortSignal }
      );
    }
    if (!this.config.credentials) {
      throw new Error(
        "Nova-3 transcription via REST requires credentials in the config. This is a bug \u2014 credentials should have been set by createWorkersAI."
      );
    }
    return createRunBinary(
      this.config.credentials,
      this.modelId,
      audioBytes,
      mediaType,
      abortSignal
    );
  }
  normalizeNova3Response(raw, warnings) {
    const results = raw.results;
    const channels = results?.channels;
    const alt = channels?.[0]?.alternatives?.[0];
    const text = alt?.transcript ?? "";
    const segments = [];
    if (alt?.words && Array.isArray(alt.words)) {
      for (const w of alt.words) {
        segments.push({
          text: w.word ?? "",
          startSecond: w.start ?? 0,
          endSecond: w.end ?? 0
        });
      }
    }
    return {
      text,
      segments,
      language: void 0,
      durationInSeconds: void 0,
      warnings,
      response: {
        timestamp: /* @__PURE__ */ new Date(),
        modelId: this.modelId,
        headers: {}
      }
    };
  }
};
function uint8ArrayToBase64(bytes) {
  let binary = "";
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

// src/workersai-speech-model.ts
var WorkersAISpeechModel = class {
  constructor(modelId, settings, config) {
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
    __publicField(this, "specificationVersion", "v3");
  }
  get provider() {
    return this.config.provider;
  }
  async doGenerate(options) {
    const { text, voice, speed, abortSignal } = options;
    const warnings = [];
    if (options.instructions) {
      warnings.push({
        details: "Workers AI TTS models do not support instructions.",
        feature: "instructions",
        type: "unsupported"
      });
    }
    if (options.outputFormat) {
      warnings.push({
        details: "Workers AI TTS models do not support output format selection. Audio is returned as MP3.",
        feature: "outputFormat",
        type: "unsupported"
      });
    }
    const inputs = { text };
    if (voice) inputs.voice = voice;
    if (speed != null) inputs.speed = speed;
    const result = await this.config.binding.run(
      this.modelId,
      inputs,
      {
        gateway: this.config.gateway,
        signal: abortSignal,
        // returnRawResponse prevents the createRun REST shim from trying
        // to JSON.parse binary audio. Real env.AI bindings don't recognize
        // this option — it has no effect, and the binding returns the normal
        // binary result (Uint8Array/ReadableStream) which toUint8Array handles.
        returnRawResponse: true
      }
    );
    const audio = await toUint8Array2(result);
    return {
      audio,
      warnings,
      response: {
        timestamp: /* @__PURE__ */ new Date(),
        modelId: this.modelId,
        headers: {}
      }
    };
  }
};
async function toUint8Array2(output) {
  if (output instanceof Response) {
    return new Uint8Array(await output.arrayBuffer());
  }
  if (output instanceof Uint8Array) {
    return output;
  }
  if (output instanceof ArrayBuffer) {
    return new Uint8Array(output);
  }
  if (output instanceof ReadableStream) {
    const reader = output.getReader();
    const chunks = [];
    let totalLength = 0;
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      chunks.push(value);
      totalLength += value.length;
    }
    const result = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of chunks) {
      result.set(chunk, offset);
      offset += chunk.length;
    }
    return result;
  }
  if (typeof output === "object" && output !== null) {
    const obj = output;
    if (typeof obj.audio === "string") {
      return Uint8Array.from(atob(obj.audio), (c) => c.charCodeAt(0));
    }
  }
  throw new Error(
    `Unexpected output type from TTS model. Got ${Object.prototype.toString.call(output)}`
  );
}

// src/workersai-reranking-model.ts
var WorkersAIRerankingModel = class {
  constructor(modelId, settings, config) {
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
    __publicField(this, "specificationVersion", "v3");
  }
  get provider() {
    return this.config.provider;
  }
  async doRerank(options) {
    const { documents, query, topN, abortSignal } = options;
    const warnings = [];
    const contexts = documentsToContexts(documents, warnings);
    const inputs = {
      query,
      contexts
    };
    if (topN != null) {
      inputs.top_k = topN;
    }
    const result = await this.config.binding.run(
      this.modelId,
      inputs,
      { gateway: this.config.gateway, signal: abortSignal }
    );
    const response = result.response;
    const ranking = (response ?? []).map((item) => ({
      index: item.id ?? 0,
      relevanceScore: item.score ?? 0
    })).sort((a, b) => b.relevanceScore - a.relevanceScore);
    return {
      ranking,
      warnings,
      response: {
        timestamp: /* @__PURE__ */ new Date(),
        modelId: this.modelId,
        headers: {}
      }
    };
  }
};
function documentsToContexts(documents, warnings) {
  if (documents.type === "text") {
    return documents.values.map((text) => ({ text }));
  }
  warnings.push({
    message: "Workers AI reranker expects text contexts. JSON objects have been stringified.",
    type: "other"
  });
  return documents.values.map((obj) => ({ text: JSON.stringify(obj) }));
}

// src/autorag-chat-language-model.ts
var AutoRAGChatLanguageModel = class extends AISearchChatLanguageModel {
};

// src/index.ts
function createWorkersAI(options) {
  if (!options.binding && !("accountId" in options && "apiKey" in options)) {
    throw new Error(
      "Invalid Workers AI configuration: you must provide either a binding (e.g. { binding: env.AI }) or credentials ({ accountId, apiKey })."
    );
  }
  let binding;
  const isBinding = !!options.binding;
  if (options.binding) {
    binding = options.binding;
  } else {
    const { accountId, apiKey } = options;
    binding = {
      run: createRun({ accountId, apiKey })
    };
  }
  const createChatModel = (modelId, settings = {}) => new WorkersAIChatLanguageModel(modelId, settings, {
    binding,
    gateway: options.gateway,
    provider: "workersai.chat",
    isBinding
  });
  const createImageModel = (modelId, settings = {}) => new WorkersAIImageModel(modelId, settings, {
    binding,
    gateway: options.gateway,
    provider: "workersai.image"
  });
  const createEmbeddingModel = (modelId, settings = {}) => new WorkersAIEmbeddingModel(modelId, settings, {
    binding,
    gateway: options.gateway,
    provider: "workersai.embedding"
  });
  const createTranscriptionModel = (modelId, settings = {}) => new WorkersAITranscriptionModel(modelId, settings, {
    binding,
    gateway: options.gateway,
    provider: "workersai.transcription",
    isBinding,
    credentials: !isBinding && "accountId" in options ? { accountId: options.accountId, apiKey: options.apiKey } : void 0
  });
  const createSpeechModel = (modelId, settings = {}) => new WorkersAISpeechModel(modelId, settings, {
    binding,
    gateway: options.gateway,
    provider: "workersai.speech"
  });
  const createRerankingModel = (modelId, settings = {}) => new WorkersAIRerankingModel(modelId, settings, {
    binding,
    gateway: options.gateway,
    provider: "workersai.reranking"
  });
  const provider = (modelId, settings) => {
    if (new.target) {
      throw new Error("The WorkersAI model function cannot be called with the new keyword.");
    }
    return createChatModel(modelId, settings);
  };
  provider.chat = createChatModel;
  provider.embedding = createEmbeddingModel;
  provider.textEmbedding = createEmbeddingModel;
  provider.textEmbeddingModel = createEmbeddingModel;
  provider.image = createImageModel;
  provider.imageModel = createImageModel;
  provider.transcription = createTranscriptionModel;
  provider.transcriptionModel = createTranscriptionModel;
  provider.speech = createSpeechModel;
  provider.speechModel = createSpeechModel;
  provider.reranking = createRerankingModel;
  provider.rerankingModel = createRerankingModel;
  return provider;
}
function createAISearch(options, providerName = "aisearch.chat") {
  const binding = options.binding;
  const createChatModel = (settings = {}) => new AISearchChatLanguageModel("@cf/meta/llama-3.3-70b-instruct-fp8-fast", settings, {
    binding,
    provider: providerName
  });
  const provider = (settings) => {
    if (new.target) {
      throw new Error("The AISearch model function cannot be called with the new keyword.");
    }
    return createChatModel(settings);
  };
  provider.chat = createChatModel;
  return provider;
}
var autoRAGWarned = false;
function createAutoRAG(options) {
  if (!autoRAGWarned) {
    autoRAGWarned = true;
    console.warn(
      "[workers-ai-provider] createAutoRAG is deprecated. Use createAISearch instead. AutoRAG has been renamed to AI Search. See https://developers.cloudflare.com/ai-search/"
    );
  }
  return createAISearch(options, "autorag.chat");
}
export {
  AISearchChatLanguageModel,
  AutoRAGChatLanguageModel,
  WorkersAIRerankingModel,
  WorkersAISpeechModel,
  WorkersAITranscriptionModel,
  createAISearch,
  createAutoRAG,
  createWorkersAI
};
