// @cloudflare/ai-chat/dist/index.js
import { MessageType } from "./types.js";
import { autoTransformMessages } from "./ai-chat-v5-migration.js";
import { t as applyChunkToParts } from "./message-builder-Bwc_Umye.js";
import { jsonSchema, tool } from "ai";
import { Agent, __DO_NOT_USE_WILL_BREAK__agentContext } from "agents";
import { nanoid } from "nanoid";

//#region src/resumable-stream.ts
/**
* ResumableStream: Standalone class for buffering, persisting, and replaying
* stream chunks in SQLite. Extracted from AIChatAgent to separate concerns.
*
* Handles:
* - Chunk buffering (batched writes to SQLite for performance)
* - Stream lifecycle (start, complete, error)
* - Chunk replay for reconnecting clients
* - Stale stream cleanup
* - Active stream restoration after agent restart
*/
/** Number of chunks to buffer before flushing to SQLite */
const CHUNK_BUFFER_SIZE = 10;
/** Maximum buffer size to prevent memory issues on rapid reconnections */
const CHUNK_BUFFER_MAX_SIZE = 100;
/** Maximum age for a "streaming" stream before considering it stale (ms) - 5 minutes */
const STREAM_STALE_THRESHOLD_MS = 300 * 1e3;
/** Default cleanup interval for old streams (ms) - every 10 minutes */
const CLEANUP_INTERVAL_MS = 600 * 1e3;
/** Default age threshold for cleaning up completed streams (ms) - 24 hours */
const CLEANUP_AGE_THRESHOLD_MS = 1440 * 60 * 1e3;
/** Shared encoder for UTF-8 byte length measurement */
const textEncoder$1 = new TextEncoder();
var ResumableStream = class ResumableStream {
	constructor(sql) {
		this.sql = sql;
		this._activeStreamId = null;
		this._activeRequestId = null;
		this._streamChunkIndex = 0;
		this._chunkBuffer = [];
		this._isFlushingChunks = false;
		this._lastCleanupTime = 0;
		this.sql`create table if not exists cf_ai_chat_stream_chunks (
      id text primary key,
      stream_id text not null,
      body text not null,
      chunk_index integer not null,
      created_at integer not null
    )`;
		this.sql`create table if not exists cf_ai_chat_stream_metadata (
      id text primary key,
      request_id text not null,
      status text not null,
      created_at integer not null,
      completed_at integer
    )`;
		this.sql`create index if not exists idx_stream_chunks_stream_id 
      on cf_ai_chat_stream_chunks(stream_id, chunk_index)`;
		this.restore();
	}
	get activeStreamId() {
		return this._activeStreamId;
	}
	get activeRequestId() {
		return this._activeRequestId;
	}
	hasActiveStream() {
		return this._activeStreamId !== null;
	}
	/**
	* Start tracking a new stream for resumable streaming.
	* Creates metadata entry in SQLite and sets up tracking state.
	* @param requestId - The unique ID of the chat request
	* @returns The generated stream ID
	*/
	start(requestId) {
		this.flushBuffer();
		const streamId = nanoid();
		this._activeStreamId = streamId;
		this._activeRequestId = requestId;
		this._streamChunkIndex = 0;
		this.sql`
      insert into cf_ai_chat_stream_metadata (id, request_id, status, created_at)
      values (${streamId}, ${requestId}, 'streaming', ${Date.now()})
    `;
		return streamId;
	}
	/**
	* Mark a stream as completed and flush any pending chunks.
	* @param streamId - The stream to mark as completed
	*/
	complete(streamId) {
		this.flushBuffer();
		this.sql`
      update cf_ai_chat_stream_metadata 
      set status = 'completed', completed_at = ${Date.now()} 
      where id = ${streamId}
    `;
		this._activeStreamId = null;
		this._activeRequestId = null;
		this._streamChunkIndex = 0;
		this._maybeCleanupOldStreams();
	}
	/**
	* Mark a stream as errored and clean up state.
	* @param streamId - The stream to mark as errored
	*/
	markError(streamId) {
		this.flushBuffer();
		this.sql`
      update cf_ai_chat_stream_metadata 
      set status = 'error', completed_at = ${Date.now()} 
      where id = ${streamId}
    `;
		this._activeStreamId = null;
		this._activeRequestId = null;
		this._streamChunkIndex = 0;
	}
	static {
		this.CHUNK_MAX_BYTES = 18e5;
	}
	/**
	* Buffer a stream chunk for batch write to SQLite.
	* Chunks exceeding the row size limit are skipped to prevent crashes.
	* The chunk is still broadcast to live clients (caller handles that),
	* but will be missing from replay on reconnection.
	* @param streamId - The stream this chunk belongs to
	* @param body - The serialized chunk body
	*/
	storeChunk(streamId, body) {
		const bodyBytes = textEncoder$1.encode(body).byteLength;
		if (bodyBytes > ResumableStream.CHUNK_MAX_BYTES) {
			console.warn(`[ResumableStream] Skipping oversized chunk (${bodyBytes} bytes) to prevent SQLite row limit crash. Live clients still receive it.`);
			return;
		}
		if (this._chunkBuffer.length >= CHUNK_BUFFER_MAX_SIZE) this.flushBuffer();
		this._chunkBuffer.push({
			id: nanoid(),
			streamId,
			body,
			index: this._streamChunkIndex
		});
		this._streamChunkIndex++;
		if (this._chunkBuffer.length >= CHUNK_BUFFER_SIZE) this.flushBuffer();
	}
	/**
	* Flush buffered chunks to SQLite in a single batch.
	* Uses a lock to prevent concurrent flush operations.
	*/
	flushBuffer() {
		if (this._isFlushingChunks || this._chunkBuffer.length === 0) return;
		this._isFlushingChunks = true;
		try {
			const chunks = this._chunkBuffer;
			this._chunkBuffer = [];
			const now = Date.now();
			for (const chunk of chunks) this.sql`
          insert into cf_ai_chat_stream_chunks (id, stream_id, body, chunk_index, created_at)
          values (${chunk.id}, ${chunk.streamId}, ${chunk.body}, ${chunk.index}, ${now})
        `;
		} finally {
			this._isFlushingChunks = false;
		}
	}
	/**
	* Send stored stream chunks to a connection for replay.
	* Chunks are marked with replay: true so the client can batch-apply them.
	* @param connection - The WebSocket connection
	* @param requestId - The original request ID
	*/
	replayChunks(connection, requestId) {
		const streamId = this._activeStreamId;
		if (!streamId) return;
		this.flushBuffer();
		const chunks = this.sql`
      select * from cf_ai_chat_stream_chunks 
      where stream_id = ${streamId} 
      order by chunk_index asc
    `;
		for (const chunk of chunks || []) connection.send(JSON.stringify({
			body: chunk.body,
			done: false,
			id: requestId,
			type: MessageType.CF_AGENT_USE_CHAT_RESPONSE,
			replay: true
		}));
		if (this._activeStreamId !== streamId) connection.send(JSON.stringify({
			body: "",
			done: true,
			id: requestId,
			type: MessageType.CF_AGENT_USE_CHAT_RESPONSE,
			replay: true
		}));
	}
	/**
	* Restore active stream state if the agent was restarted during streaming.
	* Validates stream freshness to avoid sending stale resume notifications.
	*/
	restore() {
		const activeStreams = this.sql`
      select * from cf_ai_chat_stream_metadata 
      where status = 'streaming' 
      order by created_at desc 
      limit 1
    `;
		if (activeStreams && activeStreams.length > 0) {
			const stream = activeStreams[0];
			const streamAge = Date.now() - stream.created_at;
			if (streamAge > STREAM_STALE_THRESHOLD_MS) {
				this.sql`delete from cf_ai_chat_stream_chunks where stream_id = ${stream.id}`;
				this.sql`delete from cf_ai_chat_stream_metadata where id = ${stream.id}`;
				console.warn(`[ResumableStream] Deleted stale stream ${stream.id} (age: ${Math.round(streamAge / 1e3)}s)`);
				return;
			}
			this._activeStreamId = stream.id;
			this._activeRequestId = stream.request_id;
			const lastChunk = this.sql`
        select max(chunk_index) as max_index 
        from cf_ai_chat_stream_chunks 
        where stream_id = ${this._activeStreamId}
      `;
			this._streamChunkIndex = lastChunk && lastChunk[0]?.max_index != null ? lastChunk[0].max_index + 1 : 0;
		}
	}
	/**
	* Clear all stream data (called on chat history clear).
	*/
	clearAll() {
		this._chunkBuffer = [];
		this.sql`delete from cf_ai_chat_stream_chunks`;
		this.sql`delete from cf_ai_chat_stream_metadata`;
		this._activeStreamId = null;
		this._activeRequestId = null;
		this._streamChunkIndex = 0;
	}
	/**
	* Drop all stream tables (called on destroy).
	*/
	destroy() {
		this.flushBuffer();
		this.sql`drop table if exists cf_ai_chat_stream_chunks`;
		this.sql`drop table if exists cf_ai_chat_stream_metadata`;
		this._activeStreamId = null;
		this._activeRequestId = null;
	}
	_maybeCleanupOldStreams() {
		const now = Date.now();
		if (now - this._lastCleanupTime < CLEANUP_INTERVAL_MS) return;
		this._lastCleanupTime = now;
		const cutoff = now - CLEANUP_AGE_THRESHOLD_MS;
		this.sql`
      delete from cf_ai_chat_stream_chunks 
      where stream_id in (
        select id from cf_ai_chat_stream_metadata 
        where status in ('completed', 'error') and completed_at < ${cutoff}
      )
    `;
		this.sql`
      delete from cf_ai_chat_stream_metadata 
      where status in ('completed', 'error') and completed_at < ${cutoff}
    `;
	}
	/** @internal For testing only */
	getStreamChunks(streamId) {
		return this.sql`
        select body, chunk_index from cf_ai_chat_stream_chunks 
        where stream_id = ${streamId} 
        order by chunk_index asc
      ` || [];
	}
	/** @internal For testing only */
	getStreamMetadata(streamId) {
		const result = this.sql`
      select status, request_id from cf_ai_chat_stream_metadata 
      where id = ${streamId}
    `;
		return result && result.length > 0 ? result[0] : null;
	}
	/** @internal For testing only */
	getAllStreamMetadata() {
		return this.sql`select id, status, request_id, created_at from cf_ai_chat_stream_metadata` || [];
	}
	/** @internal For testing only */
	insertStaleStream(streamId, requestId, ageMs) {
		const createdAt = Date.now() - ageMs;
		this.sql`
      insert into cf_ai_chat_stream_metadata (id, request_id, status, created_at)
      values (${streamId}, ${requestId}, 'streaming', ${createdAt})
    `;
	}
};

//#endregion
//#region src/index.ts
/** Shared encoder for UTF-8 byte length measurement */
const textEncoder = new TextEncoder();
/**
* Validates that a parsed message has the minimum required structure.
* Returns false for messages that would cause runtime errors downstream
* (e.g. in convertToModelMessages or the UI layer).
*
* Checks:
* - `id` is a non-empty string
* - `role` is one of the valid roles
* - `parts` is an array (may be empty — the AI SDK enforces nonempty
*   on incoming messages, but we are lenient on persisted data)
*/
function isValidMessageStructure(msg) {
	if (typeof msg !== "object" || msg === null) return false;
	const m = msg;
	if (typeof m.id !== "string" || m.id.length === 0) return false;
	if (m.role !== "user" && m.role !== "assistant" && m.role !== "system") return false;
	if (!Array.isArray(m.parts)) return false;
	return true;
}
/**
* One-shot deprecation warnings (warns once per key per session).
*/
const _deprecationWarnings = /* @__PURE__ */ new Set();
function warnDeprecated(id, message) {
	if (!_deprecationWarnings.has(id)) {
		_deprecationWarnings.add(id);
		console.warn(`[@cloudflare/ai-chat] Deprecated: ${message}`);
	}
}
/**
* Converts client tool schemas to AI SDK tool format.
*
* These tools have no `execute` function - when the AI model calls them,
* the tool call is sent back to the client for execution.
*
* @param clientTools - Array of tool schemas from the client
* @returns Record of AI SDK tools that can be spread into your tools object
*
* @deprecated Define tools on the server using `tool()` from "ai" instead.
* For tools that need client-side execution, omit the `execute` function
* and handle them via the `onToolCall` callback in `useAgentChat`.
*
* @example
* ```typescript
* // Server: Define tool without execute
* const tools = {
*   getLocation: tool({
*     description: "Get user's location",
*     inputSchema: z.object({})
*     // No execute = client must handle
*   })
* };
*
* // Client: Handle in onToolCall
* useAgentChat({
*   onToolCall: async ({ toolCall, addToolOutput }) => {
*     if (toolCall.toolName === 'getLocation') {
*       const pos = await navigator.geolocation.getCurrentPosition();
*       addToolOutput({ toolCallId: toolCall.toolCallId, output: pos });
*     }
*   }
* });
* ```
*/
function createToolsFromClientSchemas(clientTools) {
	warnDeprecated("createToolsFromClientSchemas", "createToolsFromClientSchemas() is deprecated. Define tools on the server using tool() from 'ai' and handle client execution via onToolCall in useAgentChat. Will be removed in the next major version.");
	if (!clientTools || clientTools.length === 0) return {};
	const seenNames = /* @__PURE__ */ new Set();
	for (const t of clientTools) {
		if (seenNames.has(t.name)) console.warn(`[createToolsFromClientSchemas] Duplicate tool name "${t.name}" found. Later definitions will override earlier ones.`);
		seenNames.add(t.name);
	}
	return Object.fromEntries(clientTools.map((t) => [t.name, tool({
		description: t.description ?? "",
		inputSchema: jsonSchema(t.parameters ?? { type: "object" })
	})]));
}
const decoder = new TextDecoder();
/**
* Extension of Agent with built-in chat capabilities
* @template Env Environment type containing bindings
*/
var AIChatAgent = class AIChatAgent extends Agent {
	static {
		this.ROW_MAX_BYTES = 18e5;
	}
	/** Measure UTF-8 byte length of a string (accurate for SQLite row limits). */
	static _byteLength(s) {
		return textEncoder.encode(s).byteLength;
	}
	constructor(ctx, env) {
		super(ctx, env);
		this._streamingMessage = null;
		this._approvalPersistedMessageId = null;
		this._streamCompletionPromise = null;
		this._streamCompletionResolve = null;
		this._pendingResumeConnections = /* @__PURE__ */ new Set();
		this._persistedMessageCache = /* @__PURE__ */ new Map();
		this.maxPersistedMessages = void 0;
		this.sql`create table if not exists cf_ai_chat_agent_messages (
      id text primary key,
      message text not null,
      created_at datetime default current_timestamp
    )`;
		this.sql`create table if not exists cf_ai_chat_request_context (
      key text primary key,
      value text not null
    )`;
		this._restoreRequestContext();
		this._resumableStream = new ResumableStream(this.sql.bind(this));
		this.messages = autoTransformMessages(this._loadMessagesFromDb());
		this._chatMessageAbortControllers = /* @__PURE__ */ new Map();
		const _onConnect = this.onConnect.bind(this);
		this.onConnect = async (connection, ctx) => {
			if (this._resumableStream.hasActiveStream()) this._notifyStreamResuming(connection);
			return _onConnect(connection, ctx);
		};
		const _onClose = this.onClose.bind(this);
		this.onClose = async (connection, code, reason, wasClean) => {
			this._pendingResumeConnections.delete(connection.id);
			return _onClose(connection, code, reason, wasClean);
		};
		const _onMessage = this.onMessage.bind(this);
		this.onMessage = async (connection, message) => {
			if (typeof message === "string") {
				let data;
				try {
					data = JSON.parse(message);
				} catch (_error) {
					return _onMessage(connection, message);
				}
				if (data.type === MessageType.CF_AGENT_USE_CHAT_REQUEST && data.init.method === "POST") {
					const { body } = data.init;
					if (!body) {
						console.warn("[AIChatAgent] Received chat request with empty body, ignoring");
						return;
					}
					let parsed;
					try {
						parsed = JSON.parse(body);
					} catch (_parseError) {
						console.warn("[AIChatAgent] Received chat request with invalid JSON body, ignoring");
						return;
					}
					const { messages, clientTools, ...customBody } = parsed;
					this._lastClientTools = clientTools?.length ? clientTools : void 0;
					this._lastBody = Object.keys(customBody).length > 0 ? customBody : void 0;
					this._persistRequestContext();
					const transformedMessages = autoTransformMessages(messages);
					this._broadcastChatMessage({
						messages: transformedMessages,
						type: MessageType.CF_AGENT_CHAT_MESSAGES
					}, [connection.id]);
					await this.persistMessages(transformedMessages, [connection.id]);
					this.observability?.emit({
						displayMessage: "Chat message request",
						id: data.id,
						payload: {},
						timestamp: Date.now(),
						type: "message:request"
					}, this.ctx);
					const chatMessageId = data.id;
					const abortSignal = this._getAbortSignal(chatMessageId);
					return this._tryCatchChat(async () => {
						return __DO_NOT_USE_WILL_BREAK__agentContext.run({
							agent: this,
							connection,
							request: void 0,
							email: void 0
						}, async () => {
							const response = await this.onChatMessage(async (_finishResult) => {}, {
								abortSignal,
								clientTools,
								body: this._lastBody
							});
							if (response) await this._reply(data.id, response, [connection.id], { chatMessageId });
							else {
								console.warn(`[AIChatAgent] onChatMessage returned no response for chatMessageId: ${chatMessageId}`);
								this._broadcastChatMessage({
									body: "No response was generated by the agent.",
									done: true,
									id: data.id,
									type: MessageType.CF_AGENT_USE_CHAT_RESPONSE
								}, [connection.id]);
							}
						});
					});
				}
				if (data.type === MessageType.CF_AGENT_CHAT_CLEAR) {
					this._destroyAbortControllers();
					this.sql`delete from cf_ai_chat_agent_messages`;
					this._resumableStream.clearAll();
					this._pendingResumeConnections.clear();
					this._lastClientTools = void 0;
					this._lastBody = void 0;
					this._persistRequestContext();
					this._persistedMessageCache.clear();
					this.messages = [];
					this._broadcastChatMessage({ type: MessageType.CF_AGENT_CHAT_CLEAR }, [connection.id]);
					return;
				}
				if (data.type === MessageType.CF_AGENT_CHAT_MESSAGES) {
					const transformedMessages = autoTransformMessages(data.messages);
					await this.persistMessages(transformedMessages, [connection.id]);
					return;
				}
				if (data.type === MessageType.CF_AGENT_CHAT_REQUEST_CANCEL) {
					this._cancelChatRequest(data.id);
					return;
				}
				if (data.type === MessageType.CF_AGENT_STREAM_RESUME_REQUEST) {
					if (this._resumableStream.hasActiveStream()) this._notifyStreamResuming(connection);
					return;
				}
				if (data.type === MessageType.CF_AGENT_STREAM_RESUME_ACK) {
					this._pendingResumeConnections.delete(connection.id);
					if (this._resumableStream.hasActiveStream() && this._resumableStream.activeRequestId === data.id) this._resumableStream.replayChunks(connection, this._resumableStream.activeRequestId);
					return;
				}
				if (data.type === MessageType.CF_AGENT_TOOL_RESULT) {
					const { toolCallId, toolName, output, autoContinue, clientTools } = data;
					if (clientTools?.length) {
						this._lastClientTools = clientTools;
						this._persistRequestContext();
					}
					this._applyToolResult(toolCallId, toolName, output).then((applied) => {
						if (applied && autoContinue) {
							const waitForStream = async () => {
								if (this._streamCompletionPromise) await this._streamCompletionPromise;
								else await new Promise((resolve) => setTimeout(resolve, 500));
							};
							waitForStream().then(() => {
								const continuationId = nanoid();
								const abortSignal = this._getAbortSignal(continuationId);
								return this._tryCatchChat(async () => {
									return __DO_NOT_USE_WILL_BREAK__agentContext.run({
										agent: this,
										connection,
										request: void 0,
										email: void 0
									}, async () => {
										const response = await this.onChatMessage(async (_finishResult) => {}, {
											abortSignal,
											clientTools: clientTools ?? this._lastClientTools,
											body: this._lastBody
										});
										if (response) await this._reply(continuationId, response, [], {
											continuation: true,
											chatMessageId: continuationId
										});
									});
								});
							}).catch((error) => {
								console.error("[AIChatAgent] Tool continuation failed:", error);
							});
						}
					});
					return;
				}
				if (data.type === MessageType.CF_AGENT_TOOL_APPROVAL) {
					const { toolCallId, approved, autoContinue } = data;
					this._applyToolApproval(toolCallId, approved).then((applied) => {
						if (applied && approved && autoContinue) {
							const waitForStream = async () => {
								if (this._streamCompletionPromise) await this._streamCompletionPromise;
								else await new Promise((resolve) => setTimeout(resolve, 500));
							};
							waitForStream().then(() => {
								const continuationId = nanoid();
								const abortSignal = this._getAbortSignal(continuationId);
								return this._tryCatchChat(async () => {
									return __DO_NOT_USE_WILL_BREAK__agentContext.run({
										agent: this,
										connection,
										request: void 0,
										email: void 0
									}, async () => {
										const response = await this.onChatMessage(async (_finishResult) => {}, {
											abortSignal,
											clientTools: this._lastClientTools,
											body: this._lastBody
										});
										if (response) await this._reply(continuationId, response, [], {
											continuation: true,
											chatMessageId: continuationId
										});
									});
								});
							}).catch((error) => {
								console.error("[AIChatAgent] Tool approval continuation failed:", error);
							});
						}
					});
					return;
				}
			}
			return _onMessage(connection, message);
		};
	}
	/**
	* Notify a connection about an active stream that can be resumed.
	* The client should respond with CF_AGENT_STREAM_RESUME_ACK to receive chunks.
	* @param connection - The WebSocket connection to notify
	*/
	_notifyStreamResuming(connection) {
		if (!this._resumableStream.hasActiveStream()) return;
		this._pendingResumeConnections.add(connection.id);
		connection.send(JSON.stringify({
			type: MessageType.CF_AGENT_STREAM_RESUMING,
			id: this._resumableStream.activeRequestId
		}));
	}
	/** @internal Delegate to _resumableStream */
	get _activeStreamId() {
		return this._resumableStream?.activeStreamId ?? null;
	}
	/** @internal Delegate to _resumableStream */
	get _activeRequestId() {
		return this._resumableStream?.activeRequestId ?? null;
	}
	/** @internal Delegate to _resumableStream */
	_startStream(requestId) {
		return this._resumableStream.start(requestId);
	}
	/** @internal Delegate to _resumableStream */
	_completeStream(streamId) {
		this._resumableStream.complete(streamId);
		this._pendingResumeConnections.clear();
	}
	/** @internal Delegate to _resumableStream */
	_storeStreamChunk(streamId, body) {
		this._resumableStream.storeChunk(streamId, body);
	}
	/** @internal Delegate to _resumableStream */
	_flushChunkBuffer() {
		this._resumableStream.flushBuffer();
	}
	/** @internal Delegate to _resumableStream */
	_restoreActiveStream() {
		this._resumableStream.restore();
	}
	/** @internal Delegate to _resumableStream */
	_markStreamError(streamId) {
		this._resumableStream.markError(streamId);
	}
	/**
	* Restore _lastBody and _lastClientTools from SQLite.
	* Called in the constructor so these values survive DO hibernation.
	* @internal
	*/
	_restoreRequestContext() {
		const rows = this.sql`
        select key, value from cf_ai_chat_request_context
      ` || [];
		for (const row of rows) try {
			if (row.key === "lastBody") this._lastBody = JSON.parse(row.value);
			else if (row.key === "lastClientTools") this._lastClientTools = JSON.parse(row.value);
		} catch {}
	}
	/**
	* Persist _lastBody and _lastClientTools to SQLite so they survive hibernation.
	* Uses upsert (INSERT OR REPLACE) so repeated calls are safe.
	* @internal
	*/
	_persistRequestContext() {
		if (this._lastBody) this.sql`
        insert or replace into cf_ai_chat_request_context (key, value)
        values ('lastBody', ${JSON.stringify(this._lastBody)})
      `;
		else this.sql`delete from cf_ai_chat_request_context where key = 'lastBody'`;
		if (this._lastClientTools) this.sql`
        insert or replace into cf_ai_chat_request_context (key, value)
        values ('lastClientTools', ${JSON.stringify(this._lastClientTools)})
      `;
		else this.sql`delete from cf_ai_chat_request_context where key = 'lastClientTools'`;
	}
	_broadcastChatMessage(message, exclude) {
		const allExclusions = [...exclude || [], ...this._pendingResumeConnections];
		this.broadcast(JSON.stringify(message), allExclusions);
	}
	/**
	* Broadcasts a text event for non-SSE responses.
	* This ensures plain text responses follow the AI SDK v5 stream protocol.
	*
	* @param streamId - The stream identifier for chunk storage
	* @param event - The text event payload (text-start, text-delta with delta, or text-end)
	* @param continuation - Whether this is a continuation of a previous stream
	*/
	_broadcastTextEvent(streamId, event, continuation) {
		const body = JSON.stringify(event);
		this._storeStreamChunk(streamId, body);
		this._broadcastChatMessage({
			body,
			done: false,
			id: event.id,
			type: MessageType.CF_AGENT_USE_CHAT_RESPONSE,
			...continuation && { continuation: true }
		});
	}
	_loadMessagesFromDb() {
		const rows = this.sql`select * from cf_ai_chat_agent_messages order by created_at` || [];
		this._persistedMessageCache.clear();
		return rows.map((row) => {
			try {
				const messageStr = row.message;
				const parsed = JSON.parse(messageStr);
				if (!isValidMessageStructure(parsed)) {
					console.warn(`[AIChatAgent] Skipping invalid message ${row.id}: missing or malformed id, role, or parts`);
					return null;
				}
				this._persistedMessageCache.set(parsed.id, messageStr);
				return parsed;
			} catch (error) {
				console.error(`Failed to parse message ${row.id}:`, error);
				return null;
			}
		}).filter((msg) => msg !== null);
	}
	async onRequest(request) {
		return this._tryCatchChat(async () => {
			if (new URL(request.url).pathname.endsWith("/get-messages")) {
				const messages = this._loadMessagesFromDb();
				return Response.json(messages);
			}
			return super.onRequest(request);
		});
	}
	async _tryCatchChat(fn) {
		try {
			return await fn();
		} catch (e) {
			throw this.onError(e);
		}
	}
	/**
	* Handle incoming chat messages and generate a response
	* @param onFinish Callback to be called when the response is finished
	* @param options Options including abort signal and client-defined tools
	* @returns Response to send to the client or undefined
	*/
	async onChatMessage(onFinish, options) {
		throw new Error("received a chat message, override onChatMessage and return a Response to send to the client");
	}
	/**
	* Save messages on the server side
	* @param messages Chat messages to save
	*/
	async saveMessages(messages) {
		await this.persistMessages(messages);
		await this._tryCatchChat(async () => {
			const response = await this.onChatMessage(() => {});
			if (response) this._reply(crypto.randomUUID(), response);
		});
	}
	async persistMessages(messages, excludeBroadcastIds = []) {
		const mergedMessages = this._mergeIncomingWithServerState(messages);
		for (const message of mergedMessages) {
			const sanitizedMessage = this._sanitizeMessageForPersistence(message);
			const resolved = this._resolveMessageForToolMerge(sanitizedMessage);
			const safe = this._enforceRowSizeLimit(resolved);
			const json = JSON.stringify(safe);
			if (this._persistedMessageCache.get(safe.id) === json) continue;
			this.sql`
        insert into cf_ai_chat_agent_messages (id, message)
        values (${safe.id}, ${json})
        on conflict(id) do update set message = excluded.message
      `;
			this._persistedMessageCache.set(safe.id, json);
		}
		if (this.maxPersistedMessages != null) this._enforceMaxPersistedMessages();
		this.messages = autoTransformMessages(this._loadMessagesFromDb());
		this._broadcastChatMessage({
			messages: mergedMessages,
			type: MessageType.CF_AGENT_CHAT_MESSAGES
		}, excludeBroadcastIds);
	}
	/**
	* Merges incoming messages with existing server state.
	* This preserves tool outputs that the server has (via _applyToolResult)
	* but the client doesn't have yet.
	*
	* @param incomingMessages - Messages from the client
	* @returns Messages with server's tool outputs preserved
	*/
	_mergeIncomingWithServerState(incomingMessages) {
		const serverToolOutputs = /* @__PURE__ */ new Map();
		for (const msg of this.messages) {
			if (msg.role !== "assistant") continue;
			for (const part of msg.parts) if ("toolCallId" in part && "state" in part && part.state === "output-available" && "output" in part) serverToolOutputs.set(part.toolCallId, part.output);
		}
		if (serverToolOutputs.size === 0) return incomingMessages;
		return incomingMessages.map((msg) => {
			if (msg.role !== "assistant") return msg;
			let hasChanges = false;
			const updatedParts = msg.parts.map((part) => {
				if ("toolCallId" in part && "state" in part && part.state === "input-available" && serverToolOutputs.has(part.toolCallId)) {
					hasChanges = true;
					return {
						...part,
						state: "output-available",
						output: serverToolOutputs.get(part.toolCallId)
					};
				}
				return part;
			});
			return hasChanges ? {
				...msg,
				parts: updatedParts
			} : msg;
		});
	}
	/**
	* Resolves a message for persistence, handling tool result merging.
	* If the message contains tool parts with output-available state, checks if there's
	* an existing message with the same toolCallId that should be updated instead of
	* creating a duplicate. This prevents the "Duplicate item found" error from OpenAI
	* when client-side tool results arrive in a new request.
	*
	* @param message - The message to potentially merge
	* @returns The message with the correct ID (either original or merged)
	*/
	_resolveMessageForToolMerge(message) {
		if (message.role !== "assistant") return message;
		for (const part of message.parts) if ("toolCallId" in part && "state" in part && (part.state === "output-available" || part.state === "approval-responded" || part.state === "approval-requested")) {
			const toolCallId = part.toolCallId;
			const existingMessage = this._findMessageByToolCallId(toolCallId);
			if (existingMessage && existingMessage.id !== message.id) return {
				...message,
				id: existingMessage.id
			};
		}
		return message;
	}
	/**
	* Finds an existing assistant message that contains a tool part with the given toolCallId.
	* Used to detect when a tool result should update an existing message rather than
	* creating a new one.
	*
	* @param toolCallId - The tool call ID to search for
	* @returns The existing message if found, undefined otherwise
	*/
	_findMessageByToolCallId(toolCallId) {
		for (const msg of this.messages) {
			if (msg.role !== "assistant") continue;
			for (const part of msg.parts) if ("toolCallId" in part && part.toolCallId === toolCallId) return msg;
		}
	}
	/**
	* Sanitizes a message for persistence by removing ephemeral provider-specific
	* data that should not be stored or sent back in subsequent requests.
	*
	* This handles two issues with the OpenAI Responses API:
	*
	* 1. **Duplicate item IDs**: The AI SDK's @ai-sdk/openai provider (v2.0.x+)
	*    defaults to using OpenAI's Responses API which assigns unique itemIds
	*    to each message part. When these IDs are persisted and sent back,
	*    OpenAI rejects them as duplicates.
	*
	* 2. **Empty reasoning parts**: OpenAI may return reasoning parts with empty
	*    text and encrypted content. These cause "Non-OpenAI reasoning parts are
	*    not supported" warnings when sent back via convertToModelMessages().
	*
	* @param message - The message to sanitize
	* @returns A new message with ephemeral provider data removed
	*/
	_sanitizeMessageForPersistence(message) {
		const sanitizedParts = message.parts.filter((part) => {
			if (part.type === "reasoning") {
				const reasoningPart = part;
				if (!reasoningPart.text || reasoningPart.text.trim() === "") return false;
			}
			return true;
		}).map((part) => {
			let sanitizedPart = part;
			if ("providerMetadata" in sanitizedPart && sanitizedPart.providerMetadata && typeof sanitizedPart.providerMetadata === "object" && "openai" in sanitizedPart.providerMetadata) sanitizedPart = this._stripOpenAIMetadata(sanitizedPart, "providerMetadata");
			if ("callProviderMetadata" in sanitizedPart && sanitizedPart.callProviderMetadata && typeof sanitizedPart.callProviderMetadata === "object" && "openai" in sanitizedPart.callProviderMetadata) sanitizedPart = this._stripOpenAIMetadata(sanitizedPart, "callProviderMetadata");
			return sanitizedPart;
		});
		return {
			...message,
			parts: sanitizedParts
		};
	}
	/**
	* Helper to strip OpenAI-specific ephemeral fields from a metadata object.
	* Removes itemId and reasoningEncryptedContent while preserving other fields.
	*/
	_stripOpenAIMetadata(part, metadataKey) {
		const metadata = part[metadataKey];
		if (!metadata?.openai) return part;
		const { itemId: _itemId, reasoningEncryptedContent: _rec, ...restOpenai } = metadata.openai;
		const hasOtherOpenaiFields = Object.keys(restOpenai).length > 0;
		const { openai: _openai, ...restMetadata } = metadata;
		let newMetadata;
		if (hasOtherOpenaiFields) newMetadata = {
			...restMetadata,
			openai: restOpenai
		};
		else if (Object.keys(restMetadata).length > 0) newMetadata = restMetadata;
		const { [metadataKey]: _oldMeta, ...restPart } = part;
		if (newMetadata) return {
			...restPart,
			[metadataKey]: newMetadata
		};
		return restPart;
	}
	/**
	* Deletes oldest messages from SQLite when the count exceeds maxPersistedMessages.
	* Called after each persist to keep storage bounded.
	*/
	_enforceMaxPersistedMessages() {
		if (this.maxPersistedMessages == null) return;
		const count = this.sql`
      select count(*) as cnt from cf_ai_chat_agent_messages
    `?.[0]?.cnt ?? 0;
		if (count <= this.maxPersistedMessages) return;
		const excess = count - this.maxPersistedMessages;
		const toDelete = this.sql`
      select id from cf_ai_chat_agent_messages 
      order by created_at asc 
      limit ${excess}
    `;
		if (toDelete && toDelete.length > 0) for (const row of toDelete) {
			this.sql`delete from cf_ai_chat_agent_messages where id = ${row.id}`;
			this._persistedMessageCache.delete(row.id);
		}
	}
	/**
	* Enforces SQLite row size limits by compacting tool outputs and text parts
	* when a serialized message exceeds the safety threshold (1.8MB).
	*
	* Only fires in pathological cases (extremely large tool outputs or text).
	* Returns the message unchanged if it fits within limits.
	*
	* Compaction strategy:
	* 1. Compact tool outputs over 1KB (replace with LLM-friendly summary)
	* 2. If still too big, truncate text parts from oldest to newest
	* 3. Add metadata so clients can detect compaction
	*
	* @param message - The message to check
	* @returns The message, compacted if necessary
	*/
	_enforceRowSizeLimit(message) {
		let json = JSON.stringify(message);
		let size = AIChatAgent._byteLength(json);
		if (size <= AIChatAgent.ROW_MAX_BYTES) return message;
		if (message.role !== "assistant") {
			console.warn(`[AIChatAgent] Non-assistant message ${message.id} is ${size} bytes, exceeds row limit. Truncating text parts.`);
			return this._truncateTextParts(message);
		}
		console.warn(`[AIChatAgent] Message ${message.id} is ${size} bytes, compacting tool outputs to fit SQLite row limit`);
		const compactedToolCallIds = [];
		const compactedParts = message.parts.map((part) => {
			if ("output" in part && "toolCallId" in part && "state" in part && part.state === "output-available") {
				const outputJson = JSON.stringify(part.output);
				if (outputJson.length > 1e3) {
					compactedToolCallIds.push(part.toolCallId);
					return {
						...part,
						output: `This tool output was too large to persist in storage (${outputJson.length} bytes). If the user asks about this data, suggest re-running the tool. Preview: ${outputJson.slice(0, 500)}...`
					};
				}
			}
			return part;
		});
		let result = {
			...message,
			parts: compactedParts
		};
		if (compactedToolCallIds.length > 0) result.metadata = {
			...result.metadata ?? {},
			compactedToolOutputs: compactedToolCallIds
		};
		json = JSON.stringify(result);
		size = AIChatAgent._byteLength(json);
		if (size <= AIChatAgent.ROW_MAX_BYTES) return result;
		console.warn(`[AIChatAgent] Message ${message.id} still ${size} bytes after tool compaction, truncating text parts`);
		return this._truncateTextParts(result);
	}
	/**
	* Truncates text parts in a message to fit within the row size limit.
	* Truncates from the first text part forward, keeping the last text part
	* as intact as possible (it is usually the most relevant).
	*/
	_truncateTextParts(message) {
		const compactedTextPartIndices = [];
		const parts = [...message.parts];
		for (let i = 0; i < parts.length; i++) {
			const part = parts[i];
			if (part.type === "text" && "text" in part) {
				const text = part.text;
				if (text.length > 1e3) {
					compactedTextPartIndices.push(i);
					parts[i] = {
						...part,
						text: `[Text truncated for storage (${text.length} chars). First 500 chars: ${text.slice(0, 500)}...]`
					};
					const candidate = {
						...message,
						parts
					};
					if (AIChatAgent._byteLength(JSON.stringify(candidate)) <= AIChatAgent.ROW_MAX_BYTES) break;
				}
			}
		}
		const result = {
			...message,
			parts
		};
		if (compactedTextPartIndices.length > 0) result.metadata = {
			...result.metadata ?? {},
			compactedTextParts: compactedTextPartIndices
		};
		return result;
	}
	/**
	* Shared helper for finding a tool part by toolCallId and applying an update.
	* Handles both streaming (in-memory) and persisted (SQLite) messages.
	*
	* Checks _streamingMessage first (tool results/approvals can arrive while
	* the AI is still streaming), then retries persisted messages with backoff
	* in case streaming completes between attempts.
	*
	* @param toolCallId - The tool call ID to find
	* @param callerName - Name for log messages (e.g. "_applyToolResult")
	* @param matchStates - Which tool part states to match
	* @param applyUpdate - Mutation to apply to the matched part (streaming: in-place, persisted: spread)
	* @returns true if the update was applied, false if not found or state didn't match
	*/
	async _findAndUpdateToolPart(toolCallId, callerName, matchStates, applyUpdate) {
		let message;
		if (this._streamingMessage) {
			for (const part of this._streamingMessage.parts) if ("toolCallId" in part && part.toolCallId === toolCallId) {
				message = this._streamingMessage;
				break;
			}
		}
		if (!message) for (let attempt = 0; attempt < 10; attempt++) {
			message = this._findMessageByToolCallId(toolCallId);
			if (message) break;
			await new Promise((resolve) => setTimeout(resolve, 100));
		}
		if (!message) {
			console.warn(`[AIChatAgent] ${callerName}: Could not find message with toolCallId ${toolCallId} after retries`);
			return false;
		}
		const isStreamingMessage = message === this._streamingMessage;
		let updated = false;
		if (isStreamingMessage) {
			for (const part of message.parts) if ("toolCallId" in part && part.toolCallId === toolCallId && "state" in part && matchStates.includes(part.state)) {
				const applied = applyUpdate(part);
				Object.assign(part, applied);
				updated = true;
				break;
			}
		} else {
			const updatedParts = message.parts.map((part) => {
				if ("toolCallId" in part && part.toolCallId === toolCallId && "state" in part && matchStates.includes(part.state)) {
					updated = true;
					return applyUpdate(part);
				}
				return part;
			});
			if (updated) {
				const updatedMessage = this._sanitizeMessageForPersistence({
					...message,
					parts: updatedParts
				});
				const safe = this._enforceRowSizeLimit(updatedMessage);
				const json = JSON.stringify(safe);
				this.sql`
          update cf_ai_chat_agent_messages 
          set message = ${json}
          where id = ${message.id}
        `;
				this._persistedMessageCache.set(message.id, json);
				this.messages = autoTransformMessages(this._loadMessagesFromDb());
			}
		}
		if (!updated) {
			console.warn(`[AIChatAgent] ${callerName}: Tool part with toolCallId ${toolCallId} not in expected state (expected: ${matchStates.join("|")})`);
			return false;
		}
		if (isStreamingMessage) this._broadcastChatMessage({
			type: MessageType.CF_AGENT_MESSAGE_UPDATED,
			message
		});
		else {
			const broadcastMessage = this._findMessageByToolCallId(toolCallId);
			if (broadcastMessage) this._broadcastChatMessage({
				type: MessageType.CF_AGENT_MESSAGE_UPDATED,
				message: broadcastMessage
			});
		}
		return true;
	}
	/**
	* Applies a tool result to an existing assistant message.
	* This is used when the client sends CF_AGENT_TOOL_RESULT for client-side tools.
	* The server is the source of truth, so we update the message here and broadcast
	* the update to all clients.
	*
	* @param toolCallId - The tool call ID this result is for
	* @param _toolName - The name of the tool (unused, kept for API compat)
	* @param output - The output from the tool execution
	* @returns true if the result was applied, false if the message was not found
	*/
	async _applyToolResult(toolCallId, _toolName, output) {
		return this._findAndUpdateToolPart(toolCallId, "_applyToolResult", ["input-available"], (part) => ({
			...part,
			state: "output-available",
			output,
			preliminary: false
		}));
	}
	async _streamSSEReply(id, streamId, reader, message, streamCompleted, continuation = false) {
		streamCompleted.value = false;
		while (true) {
			const { done, value } = await reader.read();
			if (done) {
				this._completeStream(streamId);
				streamCompleted.value = true;
				this._broadcastChatMessage({
					body: "",
					done: true,
					id,
					type: MessageType.CF_AGENT_USE_CHAT_RESPONSE,
					...continuation && { continuation: true }
				});
				break;
			}
			const lines = decoder.decode(value).split("\n");
			for (const line of lines) if (line.startsWith("data: ") && line !== "data: [DONE]") try {
				const data = JSON.parse(line.slice(6));
				const handled = applyChunkToParts(message.parts, data);
				if (data.type === "tool-approval-request" && this._streamingMessage) {
					const snapshot = {
						...this._streamingMessage,
						parts: [...this._streamingMessage.parts]
					};
					const sanitized = this._sanitizeMessageForPersistence(snapshot);
					const json = JSON.stringify(sanitized);
					this.sql`
                INSERT INTO cf_ai_chat_agent_messages (id, message)
                VALUES (${sanitized.id}, ${json})
                ON CONFLICT(id) DO UPDATE SET message = excluded.message
              `;
					this._approvalPersistedMessageId = sanitized.id;
				}
				if ((data.type === "tool-output-available" || data.type === "tool-output-error") && data.toolCallId) {
					if (!message.parts.some((p) => "toolCallId" in p && p.toolCallId === data.toolCallId)) if (data.type === "tool-output-available") this._findAndUpdateToolPart(data.toolCallId, "_streamSSEReply", [
						"input-available",
						"input-streaming",
						"approval-responded",
						"approval-requested"
					], (part) => ({
						...part,
						state: "output-available",
						output: data.output,
						...data.preliminary !== void 0 && { preliminary: data.preliminary }
					}));
					else this._findAndUpdateToolPart(data.toolCallId, "_streamSSEReply", [
						"input-available",
						"input-streaming",
						"approval-responded",
						"approval-requested"
					], (part) => ({
						...part,
						state: "output-error",
						errorText: data.errorText
					}));
				}
				if (!handled) switch (data.type) {
					case "start":
						if (data.messageId != null) message.id = data.messageId;
						if (data.messageMetadata != null) message.metadata = message.metadata ? {
							...message.metadata,
							...data.messageMetadata
						} : data.messageMetadata;
						break;
					case "finish":
					case "message-metadata":
						if (data.messageMetadata != null) message.metadata = message.metadata ? {
							...message.metadata,
							...data.messageMetadata
						} : data.messageMetadata;
						break;
					case "finish-step": break;
					case "error":
						this._broadcastChatMessage({
							error: true,
							body: data.errorText ?? JSON.stringify(data),
							done: false,
							id,
							type: MessageType.CF_AGENT_USE_CHAT_RESPONSE
						});
						break;
				}
				let eventToSend = data;
				if (data.type === "finish" && "finishReason" in data) {
					const { finishReason, ...rest } = data;
					eventToSend = {
						...rest,
						type: "finish",
						messageMetadata: { finishReason }
					};
				}
				const chunkBody = JSON.stringify(eventToSend);
				this._storeStreamChunk(streamId, chunkBody);
				this._broadcastChatMessage({
					body: chunkBody,
					done: false,
					id,
					type: MessageType.CF_AGENT_USE_CHAT_RESPONSE,
					...continuation && { continuation: true }
				});
			} catch (_error) {}
		}
	}
	async _sendPlaintextReply(id, streamId, reader, message, streamCompleted, continuation = false) {
		this._broadcastTextEvent(streamId, {
			type: "text-start",
			id
		}, continuation);
		const textPart = {
			type: "text",
			text: "",
			state: "streaming"
		};
		message.parts.push(textPart);
		while (true) {
			const { done, value } = await reader.read();
			if (done) {
				textPart.state = "done";
				this._broadcastTextEvent(streamId, {
					type: "text-end",
					id
				}, continuation);
				this._completeStream(streamId);
				streamCompleted.value = true;
				this._broadcastChatMessage({
					body: "",
					done: true,
					id,
					type: MessageType.CF_AGENT_USE_CHAT_RESPONSE,
					...continuation && { continuation: true }
				});
				break;
			}
			const chunk = decoder.decode(value);
			if (chunk.length > 0) {
				textPart.text += chunk;
				this._broadcastTextEvent(streamId, {
					type: "text-delta",
					id,
					delta: chunk
				}, continuation);
			}
		}
	}
	/**
	* Applies a tool approval response from the client, updating the persisted message.
	* This is called when the client sends CF_AGENT_TOOL_APPROVAL for tools with needsApproval.
	* Updates the tool part state from input-available/approval-requested to approval-responded.
	*
	* @param toolCallId - The tool call ID this approval is for
	* @param approved - Whether the tool execution was approved
	* @returns true if the approval was applied, false if the message was not found
	*/
	async _applyToolApproval(toolCallId, approved) {
		return this._findAndUpdateToolPart(toolCallId, "_applyToolApproval", ["input-available", "approval-requested"], (part) => ({
			...part,
			state: "approval-responded",
			approval: {
				...part.approval,
				approved
			}
		}));
	}
	async _reply(id, response, excludeBroadcastIds = [], options = {}) {
		const { continuation = false, chatMessageId } = options;
		return this._tryCatchChat(async () => {
			if (!response.body) {
				this._broadcastChatMessage({
					body: "",
					done: true,
					id,
					type: MessageType.CF_AGENT_USE_CHAT_RESPONSE,
					...continuation && { continuation: true }
				});
				return;
			}
			const streamId = this._startStream(id);
			const reader = response.body.getReader();
			const message = {
				id: `assistant_${Date.now()}_${Math.random().toString(36).slice(2, 11)}`,
				role: "assistant",
				parts: []
			};
			this._streamingMessage = message;
			this._streamCompletionPromise = new Promise((resolve) => {
				this._streamCompletionResolve = resolve;
			});
			const isSSE = (response.headers.get("content-type") || "").includes("text/event-stream");
			const streamCompleted = { value: false };
			let earlyPersistedId = null;
			try {
				if (isSSE) await this._streamSSEReply(id, streamId, reader, message, streamCompleted, continuation);
				else await this._sendPlaintextReply(id, streamId, reader, message, streamCompleted, continuation);
			} catch (error) {
				if (!streamCompleted.value) {
					this._markStreamError(streamId);
					this._broadcastChatMessage({
						body: error instanceof Error ? error.message : "Stream error",
						done: true,
						error: true,
						id,
						type: MessageType.CF_AGENT_USE_CHAT_RESPONSE,
						...continuation && { continuation: true }
					});
				}
				throw error;
			} finally {
				reader.releaseLock();
				this._streamingMessage = null;
				earlyPersistedId = this._approvalPersistedMessageId;
				this._approvalPersistedMessageId = null;
				if (this._streamCompletionResolve) {
					this._streamCompletionResolve();
					this._streamCompletionResolve = null;
					this._streamCompletionPromise = null;
				}
				if (chatMessageId) {
					this._removeAbortController(chatMessageId);
					if (streamCompleted.value) this.observability?.emit({
						displayMessage: continuation ? "Chat message response (tool continuation)" : "Chat message response",
						id: chatMessageId,
						payload: {},
						timestamp: Date.now(),
						type: "message:response"
					}, this.ctx);
				}
			}
			if (message.parts.length > 0) if (earlyPersistedId) {
				const updatedMessages = this.messages.map((msg) => msg.id === earlyPersistedId ? message : msg);
				await this.persistMessages(updatedMessages, excludeBroadcastIds);
			} else if (continuation) {
				let lastAssistantIdx = -1;
				for (let i = this.messages.length - 1; i >= 0; i--) if (this.messages[i].role === "assistant") {
					lastAssistantIdx = i;
					break;
				}
				if (lastAssistantIdx >= 0) {
					const lastAssistant = this.messages[lastAssistantIdx];
					const mergedMessage = {
						...lastAssistant,
						parts: [...lastAssistant.parts, ...message.parts]
					};
					const updatedMessages = [...this.messages];
					updatedMessages[lastAssistantIdx] = mergedMessage;
					await this.persistMessages(updatedMessages, excludeBroadcastIds);
				} else await this.persistMessages([...this.messages, message], excludeBroadcastIds);
			} else await this.persistMessages([...this.messages, message], excludeBroadcastIds);
		});
	}
	/**
	* For the given message id, look up its associated AbortController
	* If the AbortController does not exist, create and store one in memory
	*
	* returns the AbortSignal associated with the AbortController
	*/
	_getAbortSignal(id) {
		if (typeof id !== "string") return;
		if (!this._chatMessageAbortControllers.has(id)) this._chatMessageAbortControllers.set(id, new AbortController());
		return this._chatMessageAbortControllers.get(id)?.signal;
	}
	/**
	* Remove an abort controller from the cache of pending message responses
	*/
	_removeAbortController(id) {
		this._chatMessageAbortControllers.delete(id);
	}
	/**
	* Propagate an abort signal for any requests associated with the given message id
	*/
	_cancelChatRequest(id) {
		if (this._chatMessageAbortControllers.has(id)) this._chatMessageAbortControllers.get(id)?.abort();
	}
	/**
	* Abort all pending requests and clear the cache of AbortControllers
	*/
	_destroyAbortControllers() {
		for (const controller of this._chatMessageAbortControllers.values()) controller?.abort();
		this._chatMessageAbortControllers.clear();
	}
	/**
	* When the DO is destroyed, cancel all pending requests and clean up resources
	*/
	async destroy() {
		this._destroyAbortControllers();
		this._resumableStream.destroy();
		await super.destroy();
	}
};

//#endregion
export { AIChatAgent, createToolsFromClientSchemas };
